src/AdvancedHMC.jl

```diff
@@ -49,6 +51,40 @@ export Trajectory, HMCKernel,
 
 # Useful defaults
 
+"""
+    $(TYPEDEF)
+
+Wrapper around something that implements the `LogDensityProblem` interface.
+
+This itself then implements the `LogDensityProblem` interface by simply deferring to the wrapped object.
+
+Since this is a sub-type of `AbstractMCMC.AbstractModel`, it is also compatible with the `AbstractMCMC` interface.
+"""
+struct LogDensityModel{L} <: AbstractMCMC.AbstractModel
+    logdensity::L
+end
+
+LogDensityModel(logdensity, ad) = LogDensityModel(ADgradient(ad, logdensity))
+
+function LogDensityProblems.ADgradient(kind::Symbol, ℓ::LogDensityModel)
+    return LogDensityModel(LogDensityProblems.ADgradient(kind, ℓ.logdensity))
+end
+
+for kind in [:ForwardDiff, :ReverseDiff, :Zygote, :Tracker, :Enzyme]
```

Comment:
A bit annoying that this is hard-coded.

Do we actually need to implement the whole LogDensityProblems interface for the model? Is that supposed to be called by users? If not, we could just use and call whatever we need without making it a complete wrapper.

Similar to how in Turing we don't define any of this for `Model` but only (and only for own AD types!) functions such as `ADgradient` for the density.

Reply:
I'm a bit uncertain what you mean by "is it supposed to be called by the user". Are you referring to `ADgradient`? If so, I would say we should allow it. IMO if we're buying into the LogDensityProblems.jl interface, we should adopt it properly so that the users can move between different packages, e.g. AdvancedHMC.jl and AdvancedMH.jl, without having to redefine their model (if they're not using a Turing model).

But I agree, it's annoying that it's hardcoded. Would even be nice if LogDensityProblems.jl just had a list of AD packages that it supports as a variable so we could do this programmatically. Or is that a bad idea? @tpapp

Reply:
1. Generally, the intention is that users only call via the `LogDensityProblems` interface when debugging or optimizing code (eg profiling, benchmarking).

2. I would be happy to add a variable with the currently supported interfaces *for use in CI*.

Reply:
I opened https://github.com/tpapp/LogDensityProblems.jl/issues/96 for discussion.

Reply:
> I would be happy to add a variable with the currently supported interfaces for use in CI.

So you're not a big fan of what I've done here? I basically want `LogDensityModel` to do nothing but propagate everything to the underlying implemented of the LogDensityProblems.jl interface. But to do this, without causing method ambiguities, I need to overload each of the `ADgradient` methods.

Reply:
My main point above was: Why does the model have to support the LogDensityProblems interface at all? When do users manually transform the model? For instance, currently in Turing we do not support these things at all for models - we only implemented them for the `Turing.LogDensityFunction`. Users specify their AD backend either globally or as part of the sampler, and then we handle everything internally.

Reply:
I think this also makes a bit of sense conceptually - AD choice is not really something model-specific but is a sampler-specific thing. And some samplers don't need to differentiate the log density function at all.

Reply:
> Generally, most samplers don't care about AD. It's mainly HMC and similar samplers, isn't it? So baking it into the model seems weird.

But it _also_ does make sense to go "the model is differentiable", no? And "baking the AD into the model" only occurs if, well, we _need_ to AD, not for _any_ model. 

> Just having a default model type that contains a log density function seems sufficient to be able to pass it around. The only requirement would be that the log density function (not the model!) supports the LogDensityProblems interface.

But this is what `LogDensityModel` from this PR achieves, no? This is literally just passing the log density function around but allowing us to also pass it through methods which require `AbstractMCMC.AbstractModel`.

Reply:
> The reason for not using just Val(:ForwardDiff) etc. in Turing but defining our own AD backends is that in that way we can collect additional options such as chunk size, tag, etc. in a clean way and provide then these information to ADgradient. It also allows us to define a backend for e.g. Zygote without actually every loading or depending on Zygote. Maybe all these things could be resolved though if LogDensityProblems/AbstractDifferentiation would support these additional options. At least in Turing we could e.g. already use AbstractDifferentiation.ForwardDiffBackend instead of Turing.ForwardDiffAD - if it would support custom tags as well. We could then just call ADgradient(Val(:ForwardDiff), ...) in the same way but without our custom backend. 

This I'm 100% behind! I'm not arguing against this:)

> These backends are hidden behind a Requires block though in AbstractDifferentiation - maybe that should and could be changed?

Indeed. This (+ IIRC, lack of support for compiled ReverseDiff) is what stopped me the last time I looked into whether AbstractDifferentiation. The AD types need to be defined prior to the Requires block so we can actually dispatch on them in downstream packages without first having to load the corresponding AD package.